// arrow-admin-api/src/endpoints/modelflows/countAPI.js
const { getShortName } = require('../util');
const { arrowQueryParameter } = require('../../util');

const countAPI = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.',
				properties: {
					where: arrowQueryParameter.where
				}
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.count',
	nodes: {
		'model.count': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'count',
			name: `Count ${getShortName(model, true)}`,
			parameters: [
				{
					name: 'where',
					value: '$.params.where'
				}
			],
			outputs: {
				next: {
					context: '$.count',
					routes: [ 'response.success' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '200',
					type: 'number'
				},
				{
					name: 'body',
					value: '$.count'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

exports = module.exports = countAPI;
