// arrow-admin-api/src/endpoints/modelflows/createAPI.js
const { getShortName, isAutoGenerated } = require('../util');

const createNoPk = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.',
				properties: {
					data: {
						$ref: schemaRef.model,
						description: `The ${model.name} to create`
					}
				},
				required: [
					'data'
				]
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.create',
	nodes: {
		'model.create': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'create',
			name: `Create ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: '$.params.data'
				}
			],
			outputs: {
				next: {
					context: '$.model',
					routes: [ 'response.success' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set success response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

const createWithPk = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.',
				properties: {
					data: {
						$ref: isAutoGenerated(model) ? schemaRef.model : schemaRef.full,
						description: `The ${model.name} to create`
					}
				},
				required: [
					'data'
				]
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.create',
	nodes: {
		'model.create': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'create',
			name: `Create ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: '$.params.data'
				}
			],
			outputs: {
				next: {
					context: '$.model',
					routes: [ 'set.headers.location' ]
				}
			}
		},
		'set.headers.location': {
			type: 'nodehandler://@axway/api-builder-plugin-fn-dot/doT',
			name: 'Set Headers',
			method: 'formatObj',
			parameters: [
				{
					name: 'data',
					type: 'jsonpath',
					value: '$.model'
				},
				{
					name: 'template',
					type: 'string',
					value: `"{\\n  \\"location\\": {{=JSON.stringify(it[\\"${model.getPrimaryKeyName()}\\"] || null)}}\\n}"`
				}
			],
			outputs: {
				next: {
					routes: [
						'response.success'
					],
					context: '$.headers'
				},
				error: {
					routes: [
						'response.error'
					],
					context: '$.error'
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set success response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'response.error': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set HTTP Response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					type: 'number',
					value: '400'
				}
			],
			outputs: {
				next: {
					context: '$.response',
					routes: []
				}
			}
		}
	}
});

const createAPI = (name, schemaRef, model) => {
	return model.hasPrimaryKey() ? createWithPk(name, schemaRef, model) : createNoPk(name, schemaRef, model);
};

exports = module.exports = createAPI;
