// arrow-admin-api/src/endpoints/modelflows/deleteAllAPI.js
const { getShortName } = require('../util');

const deleteAllAPI = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.'
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.deleteall',
	nodes: {
		'model.deleteall': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'deleteAll',
			name: `Delete all ${getShortName(model, true)}`,
			parameters: [
			],
			outputs: {
				next: {
					routes: [ 'response.success' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '204',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

exports = module.exports = deleteAllAPI;
