// arrow-admin-api/src/endpoints/modelflows/queryAPI.js
const { getShortName } = require('../util');
const { arrowQueryParameter } = require('../../util');

const queryAPI = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.',
				...arrowQueryParameter
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.query',
	nodes: {
		'model.query': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'query',
			name: `Query ${getShortName(model, true)}`,
			parameters: [
				{
					name: 'limit',
					value: '$.params.limit'
				},
				{
					name: 'skip',
					value: '$.params.skip'
				},
				{
					name: 'where',
					value: '$.params.where'
				},
				{
					name: 'order',
					value: '$.params.order'
				},
				{
					name: 'sel',
					value: '$.params.sel'
				},
				{
					name: 'unsel',
					value: '$.params.unsel'
				}
			],
			outputs: {
				next: {
					context: '$.models',
					routes: [ 'response.success' ]
				}
			}
		},
		'response.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '200',
					type: 'number'
				},
				{
					name: 'body',
					value: '$.models'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

exports = module.exports = queryAPI;
