// arrow-admin-api/src/endpoints/modelflows/upsertAPI.js
const { getShortName } = require('../util');

const upsertAPI = (name, schemaRef, model) => ({
	schemaVersion: '4',
	info: {
		name
	},
	parameter: {
		properties: {
			params: {
				type: 'object',
				description: 'The parameters from the endpoint.',
				properties: {
					data: {
						$ref: schemaRef.full,
						description: `The ${model.name} to upsert`
					}
				},
				required: [
					'data'
				],
			},
			request: {
				type: 'object',
				description: 'The HTTP request.'
			},
			config: {
				type: 'object',
				description: 'The service\'s configuration'
			},
			env: {
				type: 'object',
				description: 'The host OS environment'
			}
		},
		additionalProperties: false,
		required: [
			'params',
			'request',
			'config',
			'env'
		]
	},
	start: 'model.upsert',
	nodes: {
		'model.upsert': {
			type: `nodehandler://api-builder-flow-invoke/model/${model.name}`,
			method: 'upsert',
			name: `Upsert ${getShortName(model)}`,
			parameters: [
				{
					name: 'data',
					value: '$.params.data'
				}
			],
			outputs: {
				insert: {
					context: '$.model',
					routes: [ 'set.headers.location', 'insert.success' ]
				},
				update: {
					context: '$.model',
					routes: [ 'update.success' ]
				}
			}
		},
		'set.headers.location': {
			type: 'nodehandler://axway-flow/setcontext',
			method: 'setcontext',
			name: 'Set Location',
			parameters: [
				{
					name: 'value',
					value: '$.model.id'
				}
			],
			outputs: {
				next: {
					routes: [ 'insert.success' ],
					context: '$.headers.location'
				}
			}
		},
		'insert.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set insert response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '201',
					type: 'number'
				},
				{
					name: 'headers',
					value: '$.headers'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		},
		'update.success': {
			type: 'nodehandler://axway-flow/http',
			name: 'Set update response',
			method: 'setresponse',
			parameters: [
				{
					name: 'status',
					value: '204',
					type: 'number'
				}
			],
			outputs: {
				next: {
					context: '$.response'
				}
			}
		}
	}
});

exports = module.exports = upsertAPI;
