// arrow-admin-api/src/endpoints/modelendpoints/countAPI.js
const { arrowQueryOptions } = require('../../util');

module.exports = (flowName, operationId, schemaRef, model) => ({
	swagger: '2.0',
	info: {
		title: `${model.name}`,
		version: '1.0.0'
	},
	paths: {
		[`/${model.name.replace(/^.*?\//, '').toLowerCase()}/count`]: {
			get: {
				operationId: `${operationId}`,
				tags: [
					`${model.name}`
				],
				description: `Count a ${model.name}`,
				consumes: [],
				produces: [
					'application/json'
				],
				parameters: [ ...arrowQueryOptions.filter(opt => opt.name === 'where') ],
				responses: {
					200: {
						description: 'The count succeeded, and the results are available.',
						schema: {
							type: 'integer'
						}
					},
					400: {
						description: 'Bad request.'
					},
					401: {
						description: 'This request requires user authentication, as configured by the server.'
					},
					404: {
						description: 'No results were found.'
					},
					500: {
						description: 'Something went wrong during the request; check out the logs on your server.'
					}
				}
			}
		}
	}
});
