// arrow-admin-api/src/endpoints/modelendpoints/createAPI.js
const { isAutoGenerated } = require('../util');

module.exports = (flowName, operationId, schemaRef, model) => ({
	swagger: '2.0',
	info: {
		title: `${model.name}`,
		version: '1.0.0'
	},
	paths: {
		[`/${model.name.replace(/^.*?\//, '').toLowerCase()}`]: {
			post: {
				operationId: `${operationId}`,
				tags: [
					`${model.name}`
				],
				description: `Create a ${model.name}`,
				consumes: [
					'application/json'
				],
				produces: [
					'application/json'
				],
				parameters: [
					{
						name: 'data',
						in: 'body',
						description: 'Create body',
						schema: {
							$ref: (!model.hasPrimaryKey() || isAutoGenerated(model)) ? schemaRef.model : schemaRef.full
						}
					}
				],
				responses: {
					201: {
						description: 'The create succeeded.',
						...(model.hasPrimaryKey() ? {
							headers: {
								Location: {
									description: 'The URL to the newly created instance.',
									type: 'string'
								}
							}
						} : {})
					},
					400: {
						description: 'Bad request.'
					},
					401: {
						description: 'This request requires user authentication, as configured by the server.'
					},
					404: {
						description: 'No results were found.'
					},
					500: {
						description: 'Something went wrong during the request; check out the logs on your server.'
					}
				}
			}
		}
	}
});
