// arrow-admin-api/src/endpoints/modelendpoints/updateAPI.js

module.exports = (flowName, operationId, schemaRef, model) => ({
	swagger: '2.0',
	info: {
		title: `${model.name}`,
		version: '1.0.0'
	},
	paths: {
		[`/${model.name.replace(/^.*?\//, '').toLowerCase()}/{${model.getPrimaryKeyName()}}`]: {
			put: {
				operationId: `${operationId}`,
				tags: [
					`${model.name}`
				],
				description: `Update a ${model.name}`,
				consumes: [
					'application/json'
				],
				produces: [
					'application/json'
				],
				parameters: [
					{
						name: model.getPrimaryKeyName(),
						in: 'path',
						description: `The ${model.name} ID`,
						required: true,
						type: model.getPrimaryKeyType()
					},
					{
						name: 'data',
						in: 'body',
						description: 'Update body',
						schema: {
							$ref: schemaRef.model
						}
					}
				],
				responses: {
					204: {
						description: 'The update succeeded.'
					},
					400: {
						description: 'Bad request.'
					},
					401: {
						description: 'This request requires user authentication, as configured by the server.'
					},
					404: {
						description: 'No results were found.'
					},
					500: {
						description: 'Something went wrong during the request; check out the logs on your server.'
					}
				}
			}
		}
	}
});
