// axway-flow/lib/nodehandlers/delay.js
const FlowNode = require('../flownode');
const FlowManager = require('../flowmanager');

/**
 * Causes the flow to wait for a certain amount of time before continuing to the next node.
 * Configuration:
 *  delay: The length of time the node should wait (ms)
 */
class Delay extends FlowNode {
	delay(req, cb) {
		setTimeout(() => cb.next(), req.params.delay);
	}
}

function getDelaySpec() {
	return {
		schemaVersion: '1',
		type: FlowManager.formatNodeHandlerUri('axway-flow', 'delay'),
		name: 'Delay',
		icon: 'icon-GG-clock3',
		category: 'core',
		methods: {
			delay: {
				name: 'Delay',
				description: 'Wait for a certain amount of time before continuing to the next node.',
				parameter: {
					type: 'object',
					properties: {
						delay: {
							type: 'integer',
							description: 'The length of delay, in milliseconds.'
						}
					},
					required: [
						'delay'
					],
					additionalProperties: false
				},
				outputs: {
					next: {
						name: 'Next',
						description: 'The delay completed.',
						context: '$.delayed',
						schema: {
							type: 'integer'
						}
					}
				}
			}
		}
	};
}

exports = module.exports = {
	handler: Delay,
	spec: getDelaySpec
};
