// axway-flow/lib/nodehandlers/setcontext.js
const FlowNode = require('../flownode');
const FlowManager = require('../flowmanager');

/**
 * Output values so that they can be set on the context.
 * Example node:
 *
 *        "step.1": {
 *           "type": "setcontext",
 *           "parameters": [{
 *              "name": "value",
 *              "value": "$.foo|0"
 *           }],
 *           "outputs": {
 *               "next": {
 *                   "context": "$.bar",
 *                   "routes": [ "step.2" ]
 *           }
 *       }
 */
class SetContext extends FlowNode {
	setcontext(req, cb) {
		cb.next(null, req.params.value);
	}
}

function getSpec() {
	return {
		schemaVersion: '1',
		type: FlowManager.formatNodeHandlerUri('axway-flow', 'setcontext'),
		name: 'Set Context',
		icon: 'icon-GG-copy',
		category: 'core',
		methods: {
			setcontext: {
				name: 'Set Context',
				description: 'Echo the input parameter to the output to allow them to be set in the context.',
				parameter: {
					type: 'object',
					properties: {
						value: {}
					},
					required: [
						'value'
					],
					additionalProperties: false
				},
				outputs: {
					next: {
						name: 'Next'
					}
				}
			}
		}
	};
}

exports = module.exports = {
	handler: SetContext,
	spec: getSpec
};
